### ISSUE TITLE
TTL is not cleaning data despite setting small merge_with_ttl_timeout

### ISSUE THREAD

**Issue by tshahmuratov** [ClickHouse/ClickHouse#26955]
Maybe I'm getting this wrong but merge_with_ttl_timeout is not working as expected
1. Created table with data https://clickhouse.tech/docs/ru/sql-reference/statements/alter/ttl/ and populateded it same way
2. `alter table table_with_ttl modify setting merge_with_ttl_timeout=10;`
3. wait for 5 min just in case. Data is still there
4. Ran `optimize table table_with_ttl final;`
5. Outdated data is dropped

Is merge_with_ttl_timeout expected to run each 10 seconds and drop outdated data or we should periodically call  `optimize table table_with_ttl final` in order to do cleanup?

Clickhouse is latest version from docker image without additional settings.
Connected to ClickHouse server version 21.6.4 revision 54448.

**Comment by den-crane** [comment]
Can you share your steps? I want to reproduce the issue.

**Comment by tshahmuratov** [comment]
CREATE TABLE table_with_ttl
(
    event_time DateTime,
    UserID UInt64,
    Comment String
)
ENGINE MergeTree()
ORDER BY tuple()
TTL event_time + INTERVAL 3 MONTH;
SETTINGS min_bytes_for_wide_part = 0;

INSERT INTO table_with_ttl VALUES (now(), 1, 'username1');

INSERT INTO table_with_ttl VALUES (now() - INTERVAL 4 MONTH, 2, 'username2');

alter table table_with_ttl modify setting merge_with_ttl_timeout=10;

// wait more than 10 seconds
//data is not deleted and deleted only after optimize table

**Comment by tshahmuratov** [comment]
Actually no need obsolete version tag. Retested it on latest
Connected to ClickHouse server version 21.7.5 revision 54449.
Same behaviour

**Comment by toddyocum** [comment]
@den-crane , fwiw, I've been watching the open TTL issues, am currently hitting issues with TTL not doing it's expected behavior.

In brief, I have a MergeTree table with a TTL of 2 days, and see data from 6 days old in the table still.

I currently have 6 CH nodes in a cluster, this issue happened on 4 of the 6. 

Restarting ClickHouse fixed the issue in the sense that TTL started doing it's job again, I would only see 2 day old data in the table,  but about a week later now, I have seen at least for one of the 6 nodes , TTL is no longer working.

After looking at the code some, I think this comment in the related issue seems like a good guess of what is happening.

#24092 
[Comment 844199199 Issue 24092](https://github.com/ClickHouse/ClickHouse/issues/24092#issuecomment-844199199)

What I don't know is how the reference count for `merges_with_ttl_counter` is not being decremented always.

Since the first occurrence, I tried ALTERing my tables with the setting `ttl_only_drop_parts = 1` and  `merge_with_ttl_timeout = 3600`, with the idea that maybe a TTL operation took so long , it timed out, and never decremented the reference count. 

But given the re-occurrence, that doesn't seem to be what was happening.

So more data, fwiw.

If I do a `ALTER <table> MATERIALIZE TTL`,  the TTL operation is performed, since that is via the Mutation path vs the background merge thread pool, I think. 

I looked through the CH logs for any indications/clues of when/why/how it gets into the state, but couldn't see anything. 

I also captured a gdb backtrace while the issue was occurring,  but couldn't make much sense of it. Not sure if there is a way to examine the `merges_with_ttl_counter` directly.

I am running 21.3.4 revision 54447

BTW, I've hesitated to opening a new issue versus trying to identify an already open issue that is the same as mine, but am happy to open a new issue describing what I'm seeing if that helps the process. Not sure what the expected norm is here.

**Comment by toddyocum** [comment]
I tried attaching to the ClickHouse process to inspect the `merges_with_ttl_counter` value, but think I need to have the dbg build  of ClickHouse. 

```
(gdb) p getTotalMergesWithTTLInMergeList()
No symbol table is loaded.  Use the "file" command.
```

So I would need to install that, and see if it happens again? Not sure I can, this is on a production server, would have to check.

[merges_with_ttl_counter](https://clickhouse.tech/codebrowser/html_report/data/symbol.html?root=..///&ref=DB::MergeList::merges_with_ttl_counter)

**Comment by alexey-milovidov** [comment]
Background TTL merges do run and do drop the expired rows — no periodic `OPTIMIZE ... FINAL` is needed. Reproducing your script on 26.8, both rows in one part:

```sql
CREATE TABLE t_ttl2 (event_time DateTime, UserID UInt64)
ENGINE = MergeTree ORDER BY tuple()
TTL event_time + INTERVAL 3 MONTH
SETTINGS min_bytes_for_wide_part = 0, merge_with_ttl_timeout = 1;

INSERT INTO t_ttl2 VALUES (now(), 1), (now() - INTERVAL 4 MONTH, 2);
SELECT count() FROM t_ttl2;   -- 2, right after the insert
-- a few seconds later:
SELECT count() FROM t_ttl2;   -- 1

SELECT name, rows FROM system.parts WHERE table = 't_ttl2' AND active;
all_1_1_1	1
```

Two things about the original observation:

- `merge_with_ttl_timeout` defaults to **14400** (4 hours), and it is the *minimum interval between TTL merges of the same partition*, not a polling period that guarantees a merge within it. So waiting 5 minutes after lowering it is not enough to conclude anything on a table whose partition was last TTL-merged recently, and even with a small value the merge is scheduled by the background pool when a slot is free.
- In your two-`INSERT` script the fully-expired part is dropped at write time, so `SELECT count()` returns 1 immediately; the interesting case is the one above, where one part holds both an expired and a live row and needs a real merge.

To check what a table is waiting on, `system.parts` exposes `delete_ttl_info_min` / `delete_ttl_info_max` (and the `move_ttl_info.*`, `recompression_ttl_info.*`, `group_by_ttl_info.*`, `rows_where_ttl_info.*` nested columns) per part, and `system.part_log` records the resulting merges with `merge_reason = 'TTLDeleteMerge'`.

Closing as answered. @toddyocum's report of TTL stopping entirely until a restart is the separate `merges_with_ttl_counter` leak discussed in #24092 — please open a new issue with `system.merges` / `system.part_log` output if you still see it on a current version.
